### ISSUE TITLE
Inconsistent defaults for lower keyword in flapack.pyf.src

### ISSUE THREAD

**Issue by zerothi** [scipy/scipy#7928]
In the `flapack.pyf.src` there are many routines with upper/lower keywords to decide which part of the matrix is used for symmetric matrices.

However, while the largest part of the routines default to lower == 'U', some are defaulting to lower == 'L':

- sytf2 (https://github.com/scipy/scipy/blob/master/scipy/linalg/flapack.pyf.src#L660)
- sytrf (https://github.com/scipy/scipy/blob/master/scipy/linalg/flapack.pyf.src#L680)
- hetrf (https://github.com/scipy/scipy/blob/master/scipy/linalg/flapack.pyf.src#L884)

I think this is a mistake and should be corrected.

I will be happy to provide a PR if you agree to the change.

**Comment by pv** [comment]
Breaks backward compatibility. I think trying to change this is more 
pain than profit, so probably not worth it.

**Comment by ilayn** [comment]
sytrf and hetrf are for LDL decomposition and by definition their naming the default is lower. That's why.

**Comment by ilayn** [comment]
Speaking of which I think I should recycle the code in #6831 since we are not going past LAPACK 3.5.0 anytime soon and by the time Python 2 support is gone,  newer routines with new keywords can be added without any compatibility problems.

**Comment by zerothi** [comment]
@ilayn thanks for the clarification. However, I do not see the argument. `sytrf` can compute LDL or UDU. 
In my opinion should a routines defaults not be decided by what it is intended for when adding it. It should adhere to some kind of standard across the package. And currently all routines are defaulting the upper part.

@pv the routines have just been added in #7851, so it hasn't been released yet.

**Comment by ilayn** [comment]
There are no standards or defaults in LAPACK routines by their design only orders. Since these are low level and it goes by the name unanimously LDLt decompositions instead of Bunch/Kaufmann the natural choice is better than a consistency that doesn't exist. Note that these are the wrapper defaults not the LAPACK  defaults so there is no choice involved.

**Comment by zerothi** [comment]
Yes, there are no standards in LAPACK, however, the scipy.linalg.lapack is a package and it does implement a standard. I know that these are low level, but if I see 10 routines having lower=False by default, then I would expect it to be the default for all routines. 

Consider I wan'to to compare the factorizations from a Cholesky and a LDL decomposition, then I have to change the arguments of one of them, but not the other:

      a_po = scipy.linalg.potrf(A, lower=True)
      a_sy = scipy.linalg.sytrf(A)

**Comment by ilayn** [comment]
Yes but note that you can't use the upper part of the array and obtain an LDLt factorization. That keyword defines both the array side AND the factorization structure. It is a shortcoming of the LAPACK design which I've discussed some time ago on LAPACK board that seemingly doesn't exist anymore.

**Comment by zerothi** [comment]
I agree. But that is beyond my argument. `sytrf` may also be used in other instances where LDLt is not wanted.

So again, I think the argument of defaulting to upper is that _all other routines default to upper_. This choice, I agree, is arbitrary. But at least it is a choice.

**Comment by ilayn** [comment]
I also agree with that but what I mean is that this is also _your_ choice. Consistency among only one of the arguments across all wraps does not exist. So I think this what you would like to see instead of consistency. Because `sy/hetrf` is pretty standard for indefinite factorization and you expect to have the lower as the default (because that's how matlab/octave conditioned people). Many users won't even know the difference since 99% of the cases users supply the symmetric array. And the ones who have the triangular array will check the signature anyways.

**Comment by zerothi** [comment]
But we are in the low-level part of linear algebra routines. So people won't _just_ use it. They have to dig and implement `get_lapack_funcs` etc. to even call it. 

If this is a commonly encountered problem, then it is probably better to add a `def ldl` routine which returns postprocessed `sytrf(..,lower=True)`.

Something like this:

```
def ldl(a, overwrite_a=False, check_finite=True):
    if check_finite:
        a1 = asarray_chkfinite(a)
    else:
        a1 = asarray(a)
    if len(a1.shape) != 2:
        raise ValueError('expected matrix')
    overwrite_a = overwrite_a or (_datacopied(a1, a))
    sytrf = get_lapack_funcs('sytrf', (a1,))
    LDLt, ipiv, info = sytrf(a1, overwrite_a=overwrite_a)
    if info < 0:
        raise ValueError('illegal value in %d-th argument of '
                                            'internal lu.getrf' % -info)
    if info > 0:
        warn("Diagonal number %d is exactly zero. Singular matrix." % info,
                    RuntimeWarning)
    ipiv -= 1
    L = np.tril(LDLt, -1)
    D = np.diag(LDLt)
    return L, D, ipiv
```

@pv what do you think? Also, as I mentioned, no backwards compatibility broken, nothing has been released.

**Comment by zerothi** [comment]
Also, `sytrf` is also used for solving systems where either upper or lower parts are used. There, there are no defaults. So why not be consisent? ;)

PS. Thanks for the discussion (albeit we disagree ;) )

**Comment by ilayn** [comment]
No problem at all, I think these are quite productive discussions. The API is probably the most important part of the packages. I'm just trying to keep the intuitiveness or user-friendliness (whatever that means) of the routines because LAPACK already makes life difficult with its compact storage schemes. Thus, usage of these wrappers are probably pretty rare. That's why I assumed the position that each of them should make sense for their own right. Otherwise there will be yet another set of standards for the user to learn on top of LAPACK but depends on what we aim for.

I already prepared the `ldl` part (though it takes surprisingly much more effort to get `L` and `D` out from the array than your proposition, see #6831with ignoring the rook parts). I'll send a PR with the rook parts removed.

**Comment by zerothi** [comment]
;) Thanks! I think so too.

My position is that the flapack sources are low level and that any users would need to read the documentation to proceed using them. None of the default choices makes sense, per see. They are arbitrary. However, they _should_ be consistent. 

As such I don't think we should impose decisions on what they are intended for, regardless of popularity. They are baseline implementation that adheres to an internal decision on defaults. _Any_ usage of these functions that are _popular_ should correspondingly be created as a python method (great you are looking at a `ldl`). And hence the problem becomes that of the implementor of the `ldl` method, and will never be a problem of the user.

**Comment by pv** [comment]
If the routines haven't been in any release yet, then I don't have a 
preference, though would be nicer to be consistent. Ultimately, people 
are able to read the documentation and these low-level routines in any 
case require care to use correctly.

But it's getting late for 1.0.0, there's about two days time left, and I 
don't think we'll want to delay the release schedule for this so best to 
hurry up if you agree.

**Comment by ilayn** [comment]
OK then, @zerothi do you have time to make the PR ready soon? Because I can make it to this evening the whole wrapper and tests (they are ready anyways) and I'd like to base it on your change then to avoid commit clashes.

**Comment by zerothi** [comment]
Great @ilayn I can make it now. 

@pv I am not clear on how you do these things? Are you merging 1.0.x branch into master, or are you keeping them strictly separate?
If so, let me know and I will create a PR for 1.0.x also. See #7929 for a PR on `master`

**Comment by pv** [comment]
As a main rule, everything goes to master, any commits to be backported 
to maintenance branches are picked by the release manager.

**Comment by zerothi** [comment]
thanks for the clarification.

**Comment by rgommers** [comment]
Anything left to do here, or closed by gh-7929?

**Comment by ilayn** [comment]
Yes this is fixed in #7929
